### PR TITLE
ads: fix reconnection of envoy

### PR DESCRIPTION
**Pull request by eduser25**
Our ADS implementation had some hardcoded push update code upon
`stream.go` start that was meant to handle reconnections and ensure
a new version was always pushed for an envoy upon reconnecting.

This code suffered from several problems, among them was causing unnecessary
versioning to happen before envoy went through init stage, which caused delays
and mishandling of versions as the initial version pushed by control plane
would be ignored in all cases for new envoys.

Code now handles reconnection gracefully following protocol. A connecting envoy
that comes from some other control plane will have a version and nonce applied to it.
We can detect that situation on the control plane by checking if we ever issued
a nonce for that proxy and TypeURL. When we have no `lastNonceSent` registered and
the request comes with a non-empty one, we will:
- update `lastVersionSent` and `lastVersionApplied` to the request's version, say `X`.
- trigger an update for this TypeURL, which will generate `X+1` for this Typeurl, and
ensure the proxy has the latest config generated for this new control plane.

Signed-off-by: Eduard Serra <eduser25@gmail.com>

**Affected area**:

- Control Plane          [x]

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

**Comment by eduser25** [comment]
Example on e2e restart:

Proxy connecting that had already an applied version and a nonce from some other control plane (that died, restarted or disconnected):
```
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.cluster.v3.Cluster [nonce=1619632630691988131; version=2; resources=[]] last sent [nonce=; version=0]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type envoy.config.cluster.v3.Cluster nonce 1619632630691988131 for a proxy we didn't yet issue a nonce for. Updating version to 2."}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.listener.v3.Listener [nonce=1619632630693041937; version=2; resources=[]] last sent [nonce=; version=0]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type envoy.config.listener.v3.Listener nonce 1619632630693041937 for a proxy we didn't yet issue a nonce for. Updating version to 2."}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment [nonce=1619632630903700325; version=2; resources=[server/server]] last sent [nonce=; version=0]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type envoy.config.endpoint.v3.ClusterLoadAssignment nonce 1619632630903700325 for a proxy we didn't yet issue a nonce for. Updating version to 2."}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1619632630904534429; version=3; resources=[root-cert-for-mtls-outbound:server/server service-cert:client/client]] last sent [nonce=; version=0]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type envoy.extensions.transport_sockets.tls.v3.Secret nonce 1619632630904534429 for a proxy we didn't yet issue a nonce for. Updating version to 3."}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.route.v3.RouteConfiguration [nonce=1619632630905008232; version=2; resources=[rds-outbound]] last sent [nonce=; version=0]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type envoy.config.route.v3.RouteConfiguration nonce 1619632630905008232 for a proxy we didn't yet issue a nonce for. Updating version to 2."}
```

This situation is detected, for each type the control plane will update to the version the proxy is using (say X) and then generate version X+1
Below, ACK for all X+1 versions after the proxy connects.

```
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.cluster.v3.Cluster [nonce=1619632651917513564; version=3; resources=[]] last sent [nonce=1619632651917513564; version=3]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: ACK received for envoy.config.cluster.v3.Cluster, version: 3 nonce: 1619632651917513564 resources ACKd: Set{}"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.listener.v3.Listener [nonce=1619632651922029890; version=3; resources=[]] last sent [nonce=1619632651922029890; version=3]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: ACK received for envoy.config.listener.v3.Listener, version: 3 nonce: 1619632651922029890 resources ACKd: Set{}"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment [nonce=1619632651922800394; version=3; resources=[server/server]] last sent [nonce=1619632651922800394; version=3]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: ACK received for envoy.config.endpoint.v3.ClusterLoadAssignment, version: 3 nonce: 1619632651922800394 resources ACKd: Set{server/server}"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Secret [nonce=1619632652229842832; version=4; resources=[root-cert-for-mtls-outbound:server/server service-cert:client/client]] last sent [nonce=1619632652229842832; version=4]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: ACK received for envoy.extensions.transport_sockets.tls.v3.Secret, version: 4 nonce: 1619632652229842832 resources ACKd: Set{root-cert-for-mtls-outbound:server/server, service-cert:client/client}"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: Request type.googleapis.com/envoy.config.route.v3.RouteConfiguration [nonce=1619632652230852138; version=3; resources=[rds-outbound]] last sent [nonce=1619632652230852138; version=3]"}
PodUID=32363a85-779c-4e8f-be39-e5b633ead048: ACK received for envoy.config.route.v3.RouteConfiguration, version: 3 nonce: 1619632652230852138 resources ACKd: Set{rds-outbound}"}
```

File: pkg/envoy/ads/stream.go [hunk 1]
```diff
@@ -136,82 +138,104 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
...
 	}
 
-	// Set last version applied
-	proxy.SetLastAppliedVersion(typeURL, requestVersion)
+	// Handle NACK case
+	if discoveryRequest.ErrorDetail != nil {
+		log.Error().Msgf("Proxy SerialNumber=%s PodUID=%s: [NACK] err: \"%s\" for nonce %s, last version applied on request %s",
+			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), discoveryRequest.ErrorDetail, discoveryRequest.ResponseNonce, discoveryRequest.VersionInfo)
+		// TODO: if NACK's on our latest nonce, we can also update lastAppliedVersion
+		// TODO: if the NACK's nonce is our latest nonce, we should retry to avoid leaving the envoy in a wrong config state and update
```
Comment (shashankram) [inline pkg/envoy/ads/stream.go:198]: Are these pending issues tracked in an issue? Would be good to have issues for these.

**Review by shashankram — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]

File: pkg/envoy/ads/stream.go (hunk 1 above)
Comment (eduser25) [reply pkg/envoy/ads/stream.go:198]: https://github.com/openservicemesh/osm/issues/3254

File: pkg/envoy/ads/stream.go [hunk 2]
```diff
@@ -136,82 +138,104 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
...
+	// As per protocol, we can ignore any request on the TypeURL stream that has not caught up with last sent nonce.
 	if requestNonce != lastNonce {
 		log.Debug().Msgf("Proxy SerialNumber=%s PodUID=%s: Ignoring request for %s non-latest nonce (request: %s, current: %s)",
 			proxy.GetCertificateSerialNumber(), proxy.GetPodUID(), typeURL.Short(), requestNonce, lastNonce)
 		return false
 	}
 
+	// Nonces match
+	// At this point, there is no error and nonces match, it is guaranteed an ACK with last sent version.
+	proxy.SetLastAppliedVersion(typeURL, requestVersion)
```
Comment (draychev) [inline pkg/envoy/ads/stream.go:236]: Interesting nuance here - we set the LastAppliedVersion when nonces match and not when it is empty ✔
Which makes sense and I wonder why we had to do it the way it was done before this PR  🤔

**Review by draychev — COMMENTED** [review]
Thank you @eduser25 !!

**Pull request merged**